src/Data/SOP/Extra.hs

```diff
@@ -0,0 +1,131 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE UndecidableInstances #-}
+{-# LANGUAGE UndecidableSuperClasses #-}
+
+-- | Helpers for generics-sop
+module Data.SOP.Extra where
+
+import Data.SOP.Constraint (SListIN)
+import Data.Text qualified as T
+import GHC.TypeLits (
+  ErrorMessage (ShowType, Text, (:$$:)),
+  TypeError,
+ )
+import Generics.SOP
+import Optics.Core (Iso', iso, (%))
+import Prelude hiding (All)
+
+-- | Like `HCollapse`, but limited to 0 or 1 products in a n-ary structure.
+class HCollapseMaybe h xs where
+  hcollapseMaybe :: SListIN h xs => h (K a) xs -> Maybe a
+
+instance HCollapseMaybe NP '[] where
+  hcollapseMaybe Nil = Nothing
+
+instance HCollapseMaybe NP '[p] where
+  hcollapseMaybe (K x :* Nil) = Just x
+
+instance (ps ~ TypeError ( 'Text "Expected at most 1 product")) => HCollapseMaybe NP (p ': p1 ': ps) where
+  hcollapseMaybe _ = Nothing -- Unreachable, due to TypeError
+
+{- | `x` is contained in `xs`
+
+  Gives a free partial `Iso` between `NP I xs` and the contained element.
+
+  TODO: Can this be simplified?
+-}
+class NPContains (xs :: [Type]) (x :: Type) where
```

Comment:
Using this class is e.g. `innerRouteEncoder` means you cannot have two routes with the same model in the sum of routes, right? (they can be wrapped in a newtype, but if the user is unaware the result may be confusing before they figure it out)

Reply:
I think that's a valid concern. We need a good example that demonstrates this problem (Ex03_Store.hs uses a single model type/value throughout), before thinking about the best way to resolve it IMO.

Reply:
Yesterday I tested and indeed, if you use two routes with the same routemodel then the second one will display the model value of the first if you use `innerModel`.

Reply:
My idea for resolving this kind of issue is to use field names explicitly (just like it's done with the "interface"), like:
```hs
subRouteEncoder ::
  forall ctor field m n i o.
  HasField' field m n =>
  AsConstructor' ctor i o =>
  m ->
  RouteEncoder m i ->
  RouteEncoder n o
subRouteEncoder m =
  mapRouteEncoder equality (_Ctor' @ctor) (flip (set (field' @field)) m)
```
that can be used inside `siteOutput enc model` like `subRouteEncoder @"MySubroute" @"fieldWithModel" model enc`.

There is the issue of using it inside `siteInput`, as there you don't have the model. This is a second "problem" with `NPContains` as the current workaround is to not initialize the other models to get the "partial iso". But honestly I couldn't think of any situation using route encoder in `siteInput` that would not be better done in `siteOutput`, maybe the encoder argument should be just removed?

Reply:
[Just responding to this, while I think about the larger issue] 

> maybe the encoder argument should be just removed?

`siteInput` takes the encoder as an argument mainly to allow the user to store it in the model type. Emanote does this - so as to be able to look the encoder up from anywhere, via:

https://github.com/srid/emanote/blob/147528d9df81b881214652ce0cefec0b3d52965e/src/Emanote/Route/SiteRoute/Class.hs#L146-L148

In theory, `siteOutput` alone needs the encoder argument - but in practice, this means that we have to pass the encoder (along with the model) everywhere in Emanote.

Reply:
Okay, I _think_ I've captured the essence of the problem (with innerRouteEncoder, inneModel, etc.) in context here: https://github.com/srid/ema/issues/92#issuecomment-1163592853

Reply:
@srid 
> in practice, this means that we have to pass the encoder (along with the model) everywhere in Emanote.

Searching the `emanote` code, `siteRouteUrlStatic` and friends seem to be only used for rendering. So, it seems to me that it's really trivial to change the code adapt to a `siteInput` that don't take the `RouteEncoder` as argument: instead of your current code `emanoteSiteOutput`, just do
```
emanoteSiteOutput enc model = 
  let model' = model { modelRouteEncoder = enc }
  in render model
```

Reply:
Note that I'm only suggesting to remove the argument because this would remove the need, when defining something like `innerRouteEncoder`, for a function `InnerModel -> OuterModel`, that doesn't exist without `willNotBeUsed` "cheating". In `siteOutput` there is no problem because you can directly use the outer model (or even `\ innerModel -> outerModel { innerField = innerModel }`).

Reply:
Yup, I understand. I'm trying to remove enc from `siteInput`, but Emanote currently uses it in one place: https://github.com/srid/emanote/pull/310#issuecomment-1165676273

There probably is a way, need to think a bit more.